File: R/Regresión Lasso.R
library(readxl)
library(dplyr)
library(ggplot2)
library(glmnet)
library(tidyr)


# Leer archivo macroeconómico
df_series <- read_excel("Datos/df_series.xlsx", sheet = "Datos")

# Ruta del archivo bancario
Proyecto_4_Cuentas_de_Captación_2024 <- "Datos/Proyecto 4 Cuentas de Captación 2024.xlsx"

# Función para cargar y filtrar por fechas
cargar_filtrar <- function(sheet_name) {
  read_excel(Proyecto_4_Cuentas_de_Captación_2024, sheet = sheet_name) %>%
    mutate(Fecha = as.Date(Fecha)) %>%
    filter(Fecha >= as.Date("2019-12-01") & Fecha <= as.Date("2024-06-01"))
}

# Cargar y unir datos
TBM2 <- left_join(cargar_filtrar("TBM"), df_series, by = "Fecha")
BX2 <- left_join(cargar_filtrar("Banamex"), df_series, by = "Fecha")
BBVA2 <- left_join(cargar_filtrar("BBVA"), df_series, by = "Fecha")
SA2 <- left_join(cargar_filtrar("Santander"), df_series, by = "Fecha")
BN2 <- left_join(cargar_filtrar("Banorte"), df_series, by = "Fecha")

# Función para ajustar y analizar modelo Lasso
ajustar_y_analizar_lasso_completo <- function(df, banco, variables_respuesta, variables_predictoras) {
  cat("\n\nAnálisis completo para:", banco, "\n")
  
  df_plot_all <- data.frame(Fecha = df$Fecha)
  resultados <- list()
  
  for (var_resp in variables_respuesta) {
    cat("\n--- Variable:", var_resp, "---\n")
    
    df_filtrado <- df %>%
      dplyr::select(all_of(c("Fecha", var_resp, variables_predictoras))) %>%
      na.omit()
    
    y <- as.numeric(df_filtrado[[var_resp]])
    
    if (is.null(y) || var(y) == 0) {
      cat("Variable", var_resp, "inválida (nula o constante)\n")
      next
    }
    
    x <- scale(as.matrix(df_filtrado[, variables_predictoras]))
    
    # Ajuste Lasso
    cv_lambda <- cv.glmnet(x, y, alpha = 1)
    lambda_opt <- cv_lambda$lambda.min
    modelo <- glmnet(x, y, alpha = 1, lambda = lambda_opt)
    y_hat <- predict(modelo, newx = x)
    
    # R2
    ss_total <- sum((y - mean(y))^2)
    ss_residual <- sum((y - y_hat)^2)
    r2 <- 1 - (ss_residual / ss_total)
    
    # Spearman
    prueba_spearman <- cor.test(y, y_hat, method = "spearman")
    
    # Coeficientes
    coef_lasso <- coef(modelo)
    
    cat("R²:", round(r2, 4), "\n")
    cat("Lambda óptimo:", round(lambda_opt, 6), "\n")
    cat("Coeficientes:\n")
    print(coef_lasso)
    
    cat("Prueba Spearman:\n")
    print(prueba_spearman)
    
    cat("Interpretación económica:\n")
    for (v in variables_predictoras) {
      beta <- coef_lasso[v, 1]
      sentido <- ifelse(beta > 0, "positiva", ifelse(beta < 0, "negativa", "nula"))
      cat(v, ": relación", sentido, "\n")
    }
    
    # Agregar al dataframe de graficación
    df_nuevo <- data.frame(
      Fecha = df_filtrado$Fecha
    )
    df_nuevo[[paste0(var_resp, "_Real")]] <- y
    df_nuevo[[paste0(var_resp, "_Estimado")]] <- as.numeric(y_hat)
    
    df_plot_all <- merge(df_plot_all, df_nuevo, by = "Fecha", all.x = TRUE)
    
    # Guardar resultados
    resultados[[var_resp]] <- list(modelo = modelo, r2 = r2, coef = coef_lasso, prueba = prueba_spearman)
  }
  
  # 📈 Gráfico combinado
  df_long <- df_plot_all %>%
    pivot_longer(-Fecha, names_to = "Serie", values_to = "Valor")
  
  g <- ggplot(df_long, aes(x = Fecha, y = Valor, color = Serie, linetype = grepl("Estimado", Serie))) +
    geom_line(size = 1) +
    scale_linetype_manual(values = c("FALSE" = "solid", "TRUE" = "dashed")) +
    labs(title = paste("Modelo Lasso -", banco, "(3 variables de captación)"),
         x = "Fecha", y = "Valor", color = "Serie", linetype = "") +
    theme_minimal()
  
  print(g)
  
  return(resultados)
}




for (banco in names(bancos)) {
  df_banco <- bancos[[banco]]
  ajustar_y_analizar_lasso_completo(df_banco, banco, variables_captacion, variables_macro)
}

File: Python/Regresión Lasso.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LassoCV
from sklearn.preprocessing import StandardScaler
from scipy.stats import spearmanr
from pathlib import Path

# Leer archivo macroeconómico
df_series = pd.read_excel("Datos/df_series.xlsx", sheet_name="Datos")

# Ruta del archivo bancario
ruta_bancaria = Path("Datos/Proyecto 4 Cuentas de Captación 2024.xlsx")

# Función para cargar y filtrar por fechas
def cargar_filtrar(sheet_name):
    df = pd.read_excel(ruta_bancaria, sheet_name=sheet_name)
    df["Fecha"] = pd.to_datetime(df["Fecha"])
    return df[(df["Fecha"] >= "2019-12-01") & (df["Fecha"] <= "2024-06-01")]

# Cargar y unir datos
TBM2 = pd.merge(cargar_filtrar("TBM"), df_series, on="Fecha", how="left")
BX2 = pd.merge(cargar_filtrar("Banamex"), df_series, on="Fecha", how="left")
BBVA2 = pd.merge(cargar_filtrar("BBVA"), df_series, on="Fecha", how="left")
SA2 = pd.merge(cargar_filtrar("Santander"), df_series, on="Fecha", how="left")
BN2 = pd.merge(cargar_filtrar("Banorte"), df_series, on="Fecha", how="left")

# Función para ajustar y analizar modelo Lasso
def ajustar_y_analizar_lasso(df, banco, variable_respuesta, variables_predictoras):
    print(f"\n\nModelo Lasso: {banco} - {variable_respuesta}")

    cols = ["Fecha", variable_respuesta] + variables_predictoras
    df_filtrado = df[cols].dropna()

    if df_filtrado[variable_respuesta].nunique() <= 1:
        print(f"Variable {variable_respuesta} inválida (constante o nula)")
        return None

    X_raw = df_filtrado[variables_predictoras].values
    y = df_filtrado[variable_respuesta].values

    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X_raw)

    # ⚠️ Aumentamos el número de iteraciones para evitar problemas de convergencia
    lasso_cv = LassoCV(cv=5, max_iter=10000, n_alphas=1000).fit(X_scaled, y)
    y_pred = lasso_cv.predict(X_scaled)

    r2 = lasso_cv.score(X_scaled, y)
    print(f"\nR²: {round(r2, 4)}")
    print(f"Lambda óptimo: {round(lasso_cv.alpha_, 6)}")

    coef = pd.Series(lasso_cv.coef_, index=variables_predictoras)
    print("\nCoeficientes:")
    print(coef)

    # Prueba de Spearman
    rho, pval = spearmanr(y, y_pred)
    print("\nPrueba de Spearman:")
    print(f"rho = {rho:.4f}, p-valor = {pval:.4f}")

    # Interpretación económica
    print("\nInterpretación económica:")
    for var in variables_predictoras:
        beta = coef[var]
        if beta > 0:
            sentido = "positiva"
        elif beta < 0:
            sentido = "negativa"
        else:
            sentido = "nula"
        print(f"{var}: relación {sentido}")

    # Gráfico
    df_plot = pd.DataFrame({
        "Fecha": df_filtrado["Fecha"],
        "Real": y,
        "Estimado": y_pred
    })

    plt.figure(figsize=(10, 5))
    sns.lineplot(data=df_plot, x="Fecha", y="Real", label="Real")
    sns.lineplot(data=df_plot, x="Fecha", y="Estimado", label="Estimado", linestyle="--")
    plt.title(f"Modelo Lasso - {banco} - {variable_respuesta}")
    plt.xlabel("Fecha")
    plt.ylabel(variable_respuesta)
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.show()

    return {"modelo": lasso_cv, "r2": r2, "coef": coef, "spearman": (rho, pval)}

# Definir variables
variables_captacion = ["Depositos Vista", "Depositos Plazo", "Captación tradicional"]
variables_macro = ["TIIE", "FIX", "Base_Monetaria", "Remesas_Familiares", "INPC",
                   "Costo_Captacion", "Agregados_Monetarios", "Activos_Financieros"]

bancos = {
    "TBM": TBM2,
    "Banamex": BX2,
    "BBVA": BBVA2,
    "Santander": SA2,
    "Banorte": BN2
}

# Ejecutar el análisis Lasso para cada banco y variable
for banco, df_banco in bancos.items():
    for var_resp in variables_captacion:
        if var_resp in df_banco.columns:
            ajustar_y_analizar_lasso(df_banco, banco, var_resp, variables_macro)

